registry/storage/driver/s3-aws/s3.go

```diff
@@ -766,34 +766,59 @@ func (d *driver) Writer(ctx context.Context, path string, appendMode bool) (stor
 // Stat retrieves the FileInfo for the given path, including the current size
 // in bytes and the creation time.
 func (d *driver) Stat(ctx context.Context, path string) (storagedriver.FileInfo, error) {
-	resp, err := d.S3.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
-		Bucket:  aws.String(d.Bucket),
-		Prefix:  aws.String(d.s3Path(path)),
-		MaxKeys: aws.Int64(1),
-	})
-	if err != nil {
-		return nil, err
-	}
+	s3Path := d.s3Path(path)
 
 	fi := storagedriver.FileInfoFields{
 		Path: path,
 	}
 
-	if len(resp.Contents) == 1 {
-		if *resp.Contents[0].Key != d.s3Path(path) {
-			fi.IsDir = true
-		} else {
+	// Try HeadObject first
+	headResp, err := d.S3.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(d.Bucket),
+		Key:    aws.String(s3Path),
+	})
+	if err == nil {
```

Comment:
https://go.dev/wiki/CodeReviewComments#indent-error-flow

I would suggest refactoring this into three functions:
```go
func (d *driver) Stat(ctx context.Context, path string) (storagedriver.FileInfo, error) {
	s3Path := d.s3Path(path)
	fi, err := d.statHead(ctx, s3Path)
	if err != nil {
		if _, ok := err.(awserr.Error); ok {
			var err error
			fi, err = d.statList(ctx, s3Path)
			if err != nil {
				return nil, err
			}
		}
	}
	fi.Path = path
	return storagedriver.FileInfoInternal{FileInfoFields: fi}, nil
}
```